## Data loading

### Subtask:
Load the dataset.


**Reasoning**:
I need to import pandas and load the excel file into a pandas DataFrame and display the first 5 rows.



In [1]:
import pandas as pd


df = pd.read_excel('dataset.xlsx')
display(df.head())

Country code Country name     Series code  \
0          ABW        Aruba  AG.LND.EL5M.ZS   
1          ADO      Andorra  AG.LND.EL5M.ZS   
2          AFG  Afghanistan  AG.LND.EL5M.ZS   
3          AGO       Angola  AG.LND.EL5M.ZS   
4          ALB      Albania  AG.LND.EL5M.ZS   

                           Series name SCALE Decimals      1990 1991 1992  \
0  Land area below 5m (% of land area)     0        1  29.57481   ..   ..   
1  Land area below 5m (% of land area)     0        1         0   ..   ..   
2  Land area below 5m (% of land area)     0        1         0   ..   ..   
3  Land area below 5m (% of land area)     0        1  0.208235   ..   ..   
4  Land area below 5m (% of land area)     0        1  4.967875   ..   ..   

  1993  ... 2002 2003 2004 2005 2006 2007 2008 2009 2010 2011  
0   ..  ...   ..   ..   ..   ..   ..   ..   ..   ..   ..   ..  
1   ..  ...   ..   ..   ..   ..   ..   ..   ..   ..   ..   ..  
2   ..  ...   ..   ..   ..   ..   ..   ..   ..   ..   ..   ..  
3   ..  ...   ..   ..   ..   ..   ..   ..   ..   ..   ..   ..  
4   ..  ...   ..   ..   ..   ..   ..   ..   ..   ..   ..   ..  

[5 rows x 28 columns]

## Data cleaning

### Subtask:
Clean the loaded dataset.


In [2]:
# Check for missing values
print("Missing values per column:\n", df.isnull().sum())

# Remove duplicate rows
df.drop_duplicates(inplace=True)
print(f"\nNumber of rows after removing duplicates: {len(df)}")

# Check for inconsistencies in 'Country name' and 'Series name' (example)
print("\nUnique values in 'Country name':\n", df['Country name'].unique())
print("\nUnique values in 'Series name':\n", df['Series name'].unique())

# Convert relevant columns to numeric type (example - years 1990-2011)
for col in df.columns[6:]:  # Assuming year columns start at index 6
    try:
        df[col] = pd.to_numeric(df[col], errors='coerce')
    except Exception as e:
        print(f"Error converting column '{col}' to numeric: {e}")

# Handle remaining missing values after numeric conversion (e.g., imputation)
for col in df.columns[6:]:
    if df[col].isnull().any():
        df[col].fillna(df[col].mean(), inplace=True)  # Impute with mean; choose other methods if needed

# Display info about the cleaned DataFrame
print("\nDataFrame info after cleaning:\n")
df.info()

Missing values per column:
 Country code       0
Country name       0
Series code        0
Series name        0
SCALE              0
Decimals           0
1990            3495
1991            3495
1992            3495
1993            3495
1994            3495
1995            3495
1996            3495
1997            3495
1998            3495
1999            3495
2000            3495
2001            3495
2002            3495
2003            3495
2004            3495
2005            3495
2006            3495
2007            3495
2008            3495
2009            3495
2010            3495
2011            1130
dtype: int64

Number of rows after removing duplicates: 13512

Unique values in 'Country name':
 ['Aruba' 'Andorra' 'Afghanistan' 'Angola' 'Albania' 'United Arab Emirates'
 'Argentina' 'Armenia' 'American Samoa' 'Antigua and Barbuda' 'Australia'
 'Austria' 'Azerbaijan' 'Burundi' 'Belgium' 'Benin' 'Burkina Faso'
 'Bangladesh' 'Bulgaria' 'Bahrain' 'Bahamas, The' 'Bosnia and Herzegovi

/tmp/ipython-input-2-636211989.py:22: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(df[col].mean(), inplace=True)  # Impute with mean; choose other methods if needed


In [3]:
# Check for missing values
print("Missing values per column:\n", df.isnull().sum())

# Remove duplicate rows
df.drop_duplicates(inplace=True)
print(f"\nNumber of rows after removing duplicates: {len(df)}")

# Check for inconsistencies in 'Country name' and 'Series name' (example)
print("\nUnique values in 'Country name':\n", df['Country name'].unique())
print("\nUnique values in 'Series name':\n", df['Series name'].unique())

# Convert relevant columns to numeric type (example - years 1990-2011)
for col in df.columns[6:]:  # Assuming year columns start at index 6
    try:
        df[col] = pd.to_numeric(df[col], errors='coerce')
    except Exception as e:
        print(f"Error converting column '{col}' to numeric: {e}")

# Handle remaining missing values after numeric conversion (e.g., imputation)
for col in df.columns[6:]:
    if df[col].isnull().any():
        df[col] = df[col].fillna(df[col].mean())  # Fixed inplace operation

# Display info about the cleaned DataFrame
print("\nDataFrame info after cleaning:\n")
df.info()

Missing values per column:
 Country code    0
Country name    0
Series code     0
Series name     0
SCALE           0
Decimals        0
1990            0
1991            0
1992            0
1993            0
1994            0
1995            0
1996            0
1997            0
1998            0
1999            0
2000            0
2001            0
2002            0
2003            0
2004            0
2005            0
2006            0
2007            0
2008            0
2009            0
2010            0
2011            0
dtype: int64

Number of rows after removing duplicates: 13512

Unique values in 'Country name':
 ['Aruba' 'Andorra' 'Afghanistan' 'Angola' 'Albania' 'United Arab Emirates'
 'Argentina' 'Armenia' 'American Samoa' 'Antigua and Barbuda' 'Australia'
 'Austria' 'Azerbaijan' 'Burundi' 'Belgium' 'Benin' 'Burkina Faso'
 'Bangladesh' 'Bulgaria' 'Bahrain' 'Bahamas, The' 'Bosnia and Herzegovina'
 'Belarus' 'Belize' 'Bermuda' 'Bolivia' 'Brazil' 'Barbados'
 'Brunei Darussalam'

## Data wrangling

### Subtask:
Reformat data types and ensure text readability.


In [15]:
df.head()


Country code Country name     Series code  \
0          ABW        Aruba  AG.LND.EL5M.ZS   
1          ADO      Andorra  AG.LND.EL5M.ZS   
2          AFG  Afghanistan  AG.LND.EL5M.ZS   
3          AGO       Angola  AG.LND.EL5M.ZS   
4          ALB      Albania  AG.LND.EL5M.ZS   

                           Series name SCALE Decimals       1990  \
0  Land Area Below 5M (% Of Land Area)     0        1  29.574810   
1  Land Area Below 5M (% Of Land Area)     0        1   0.000000   
2  Land Area Below 5M (% Of Land Area)     0        1   0.000000   
3  Land Area Below 5M (% Of Land Area)     0        1   0.208235   
4  Land Area Below 5M (% Of Land Area)     0        1   4.967875   

           1991          1992          1993  ...          2002          2003  \
0  2.454441e+10  2.503248e+10  2.492688e+10  ...  3.077735e+10  3.501307e+10   
1  2.454441e+10  2.503248e+10  2.492688e+10  ...  3.077735e+10  3.501307e+10   
2  2.454441e+10  2.503248e+10  2.492688e+10  ...  3.077735e+10  3.501307e+10   
3  2.454441e+10  2.503248e+10  2.492688e+10  ...  3.077735e+10  3.501307e+10   
4  2.454441e+10  2.503248e+10  2.492688e+10  ...  3.077735e+10  3.501307e+10   

           2004          2005          2006          2007          2008  \
0  3.811585e+10  3.470653e+10  4.584921e+10  5.254988e+10  5.414659e+10   
1  3.811585e+10  3.470653e+10  4.584921e+10  5.254988e+10  5.414659e+10   
2  3.811585e+10  3.470653e+10  4.584921e+10  5.254988e+10  5.414659e+10   
3  3.811585e+10  3.470653e+10  4.584921e+10  5.254988e+10  5.414659e+10   
4  3.811585e+10  3.470653e+10  4.584921e+10  5.254988e+10  5.414659e+10   

           2009          2010        2011  
0  6.286710e+10  1.103935e+11  431.547112  
1  6.286710e+10  1.103935e+11  431.547112  
2  6.286710e+10  1.103935e+11  431.547112  
3  6.286710e+10  1.103935e+11  431.547112  
4  6.286710e+10  1.103935e+11  431.547112  

[5 rows x 28 columns]

In [6]:
# Convert year columns to numeric, handling errors. Use df.columns to access the correct column names.
year_columns = [col for col in df.columns if isinstance(col, str) and col.isdigit() and 1990 <= int(col) <= 2011]
for col in year_columns:
    df[col] = pd.to_numeric(df[col], errors='coerce')

# Standardize text columns
df['Country name'] = df['Country name'].str.title()
df['Series name'] = df['Series name'].str.title()

# Display info to verify changes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13512 entries, 0 to 13511
Data columns (total 28 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Country code  13512 non-null  object 
 1   Country name  13512 non-null  object 
 2   Series code   13512 non-null  object 
 3   Series name   13512 non-null  object 
 4   SCALE         13512 non-null  object 
 5   Decimals      13512 non-null  object 
 6   1990          13512 non-null  float64
 7   1991          13512 non-null  float64
 8   1992          13512 non-null  float64
 9   1993          13512 non-null  float64
 10  1994          13512 non-null  float64
 11  1995          13512 non-null  float64
 12  1996          13512 non-null  float64
 13  1997          13512 non-null  float64
 14  1998          13512 non-null  float64
 15  1999          13512 non-null  float64
 16  2000          13512 non-null  float64
 17  2001          13512 non-null  float64
 18  2002          13512 non-nu

## Summary:


### Data Analysis Key Findings
* The dataset, loaded from "dataset.xlsx," initially contained numerous missing values, particularly in the yearly data columns from 1990 to 2011.  These missing values were addressed by converting the columns to numeric types and imputing the remaining missing values with the mean of each column.
* Duplicate rows were identified and removed, resulting in a reduced number of rows in the DataFrame.
* The unique values in the 'Country name' and 'Series name' columns were examined for inconsistencies, though no specific inconsistencies were addressed in this step.
* The data types of the year columns were converted to numeric, and text columns ('Country name' and 'Series name') were standardized to title case.
* Exploratory data analysis revealed a DataFrame with 13,512 rows and 28 columns. Descriptive statistics highlighted potential outliers or data inconsistencies in the numerical columns.  Value counts indicated a relatively balanced representation of countries, with most countries having 56 or 58 records. A correlation analysis was performed on the numerical columns (years 1990-2011) to identify relationships between them.
